keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+  resource + operation.
+- Policy bindings are not required to match apiGroup + resource +
+  operation since the policy definition is already required to do this.
+
+This encourages policy definition authors to consider the assumptions that the
+CEL expressions make. If the expressions unconditionally access `object` without
+a `has(object)` check, the expression will only ever work on `CREATE` and
+`UPDATE` and would fail at runtime on a `DELETE`. It is quite difficult to write
+CEL expression that handle all admission requests well, so we want to guide
+policy authors toward matching only the requests that they intend to support.
+The Kubernetes admission chain also scales better, and is more resiliant, when
+matching is precise and the validation expressions don't need to do any
+post-matching checks that could have been handled by matching.
+
+Note that if a policy binding is to be applied as broadly as possible (i.e.
+everywhere allowed by the policy definition) it must do so by using a wildcard
+match rule. An `effect: MatchAll` (only permitted if no other match criteria is
+present) could be offered to make this easy, or an existing match criteria that
+supports wildcards could be used.
+
+Match Policy:
+
+`MatchPolicy` will work the same as for admission webhooks. It will default to
+`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
+definitions for different versions of CRD" for an explanation of why we need
+`MatchPolicy`.
+
+xref:
+
+- https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+- https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/
+- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#resourcepolicyrule-v1beta1-flowcontrol-apiserver-k8s-io
+- https://github.com/open-policy-agent/gatekeeper/blob/a1add93b0beb5c48eb92a6a2eb5ee7d21551a1b6/pkg/mutation/match/match.go#L22-L56
+
+#### Reporting violations to Clients
+
+This section focuses on how information is reported back to clients in
+when validations fail.
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+  - Support allow/deny result, warnings, audit annotations
+  - Support for reasons/codes
+- Ability to format message strings
+
+High level proposal:
+
+- Each policy may set a `denyReason` and/or `denyCode`
+- Each validation may define a message:
+  - `message` - may contain `{1}` format args that are supplied by
+    `meesageArgs: [<cel expression>, <cel expressiion>, ...]`
+    - If `meesage` is absent, `expression` and `name` will be included in the
+      failure message
+    - If any of the arg CEL expressions fail: `expression` and `name` will be
+      included in the failure message plus the arg evaluation failure
+- allow/deny, warnings, audit annotations will be collectively referred to as
+  `enforcement`, which will have the following levels:
+  - `Deny`,
+  - `Warn`
+  - `Audit`
```

Comment:
I was hoping for some sort of canary-ing here, like so you know if it will destroy your cluster before you turn it on.

Reply:
An orchestration layer above must create multiple policy bindings, one can be the canary.  I'm not convinced the matching criteria is sufficient for fine grained canary-ing though.

Also, we will have metrics. I suppose `enforcement: []` would need to result in metrics only for a binding, but that would be far from obvious to users.

Reply:
I was expecting canary-ing by modifying the binding object to make it more strict, which implies the need for a observe-only mode or something like that.

OTOH you could argue that the CEL policy is the thing that needs canary-ing and that kind of thing should go in the policy object.

But I think there must be some way to roll out a policy to less than 100% severity on 100% of the cluster.

Reply:
@TristonianJones and I discussed this offline. I've added a UNRESOLVED to move the responsibilities around and make the binding author responsible only for selecting between things like "Dry-Run" or "Canary" and "Enabled", and have the policy definition author responsible for what enabled means (Deny, audit various things, ...)

Reply:
per an above comment, I think if we can convince ourselves that the policy definition author has reason to want some signal as to the type of action that should be taken, we should figure out a more nuanced protocol that decouples the two parties, since the binding author *definitely* has interest in enforcement action, outside of rollout.

Take, for example, setting a policy to `Warn` not because it's a new policy, but b/c this is a dev cluster and you want to notify but not obstruct development.

Another example: forcing audit-only enforcement for critical namespaces to prioritize availability while preserving auditability.

Both of these situations are ones where a cluster admin (who I think is more likely to be writing bindings) has the necessary context, and a policy definition author (who may not even be at the same company as the cluster admin) wouldn't.